teps/0027-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,133 @@
+
+```yaml
+    finally:
+    - name: rollback
+      params:
+        - name: deployment-state
+          value: "$(context.pipelineRun.Tasks.deployment)"
+      taskSpec:
+        params:
+          - name: deployment-state
+        steps:
+          - image: ubuntu
+            name: check-deployment
+            script: |
+              if [ $(params.deployment-state) == "Failed" ]
+              then
+                echo "Deployment has failed, it's time to rollback!"
+              fi
+```
+
+The same `param` will be made available in the `dag` tasks at runtime. There might be less known use cases in which
+`dag` tasks need access to the execution status of other `dag` tasks. In general, `dag` tasks are designed using
+`runAfter`, `from`, and `task results` to build chain of dependencies, and the execution continues only if all of them
+are successful. One single failure triggers `finally` tasks which needs to be aware of such failure of `dag` task.
+
+The `dag` tasks can also be designed as guarded tasks using `when expressions`. Such guarded tasks are `skipped` if
+`when expressions` evaluate to false but `pipeline` continues executing. There might be scenarios where a `dag` task
+would like to know whether the guraded task was `skipped` or `successful`.
+
+| State | Description |
+| ----- | ----------- |
+| `Succeeded` | The `pipelineTask` was successful i.e. a respective pod was created and completed successfully. The `pipelineTask` had a `taskRun` with `Succeeded` `ConditionType` and  `True` `ConditionStatus`. |
+| `Failed` | The `pipelineTask` failed i.e. a respective pod was created but exited with error. The `pipelineTask` has a `taskRun` with `Succeeded` `ConditionType`,  `False` `ConditionStatus` and have exhausted all the retries. |
+| `Skipped` | The `pipelineTask` is `skipped` because of `when expression` or one of the parent tasks was `skipped`. It is part of `pipelineRun.Status.SkippedTasks`. |
+| `None` | This state means `none of the above`. |
+
+**None:**
+
+Within `finally`, a `pipelineTask` with this state indicates that the `pipeline` stopped executing `dag` tasks before
+it could get to this task i.e. this task was not started/executed and no execution state available.
+           
+Within `dag`, `pipelineTask` with this state is hardly possible unless a `dag` task is trying to access the execution
+state of a child task even before a parent is executed or finished executing, e.g. A -> B -> C (one branch),
+D (another branch), D is accessing execution state of B while A is running.
+
+
+### User Stories
+
+A `pipeline` author can design a `finally` task which can send success notification or failure notification depending
+on the execution status. 
```

Comment:
> (e.g. instead of "is skipped" for a task, could the task that wants that info use the same when expression?)

yes but a task without `when` expression could be skipped if one of the parents skip so same `when` expression will not be applicable for that task.